### ISSUE TITLE
Problems in using heater to simulate liquid vaporization

### ISSUE THREAD

**Issue by Jackxi1024** [IDAES/idaes-pse#400]
I'm trying to use a heater to heat N-hexane (boiling point 342.15k). When the outlet temperature does not reach the boiling point, the calculation of heat duty is OK, but when the outlet temperature exceeds the boiling point, there is a problem in the solver.
I'm using the PR equation with phase equilibrium.

**Comment by andrewlee94** [comment]
Hi @Jackxi1024 We are going to need more information to work out what is going wrong with your example. To start with, could you please provide the following:

1. The property definition you are using - without it we have no way of reproducing your error.
2. The example flowsheet you are using, so that we can make sure there are no mistakes in how you have set up the heater.
3. The full output from when you try to initialize and solve the problem. Please include as much information as possible, by setting `outlvl=idaes.logger.DEBUG` in the initialize method and  `tee=True` when using the `solve` method.

**Comment by Jackxi1024** [comment]
Thank you very much for your reply. Attached is my simulation file.

I'm trying to write a file to simulate the heater heating N-hexane (atmospheric boiling point is 69 ℃, 342.15k)
 
(1) When the outlet temperature is lower than the boiling point temperature, only the liquid phase, the calculation of heat duty is OK, which corresponds to the physical property package file hexane_property_liq.py and simulation file hexane_heater_liq.py.
 
(2) However, when the outlet temperature is higher than the boiling point temperature, the outlet is gas phase, and the result of the solver is infeasible, which corresponds to the physical property package file hexane_property_liq_vap.py and simulation file hexane_heater_liq_vap.py. Because there are two phases, I set the phase equilibrium.
 
(3) DWSIM file is a calculation method file of pure material properties based on DWSIM database. Its content and syntax are similar to RPP, but its formula is slightly different.





------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "IDAES/idaes-pse"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年6月29日(星期二) 晚上9:18
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [IDAES/idaes-pse] Problems in using heater to simulate liquid vaporization (#400)





 
Hi @Jackxi1024 We are going to need more information to work out what is going wrong with your example. To start with, could you please provide the following:
  
The property definition you are using - without it we have no way of reproducing your error.
 
The example flowsheet you are using, so that we can make sure there are no mistakes in how you have set up the heater.
 
The full output from when you try to initialize and solve the problem. Please include as much information as possible, by setting outlvl=idaes.logger.DEBUG in the initialize method and  tee=True when using the solve method.

**Comment by andrewlee94** [comment]
@Jackxi1024 Hello again,

I've taken a look at your model, and whilst I haven't worked out why it won't solve I do have some general observations.

Firstly, pure (single) component equilibrium is challenging to to in an equation oriented environment - whilst our models should be able to handle it the fact the there is a phase transition point (rather than an envelope of gradual change) make these problems more challenging.

Next, the two-phase model solves well for conditions above the bubble/dew point, which is something at least. However, if it tries to go below the bubble point then the problem becomes infeasible - I still have not worked out why though.

Finally, even though the solution is infeasible, the liquid phase enthalpy appears to be correct below the bubble point, so that calculation at least is not the problem.

In short, I don't have an answer for you yet as to why, but the issue is due to the problem being infeasible below the bubble point for some reason.

**Comment by Jackxi1024** [comment]
Hello, what puzzles me is setting the phase equilibrium. The inlet of the heater is liquid phase and the outlet is vapor phase, so there is no need to set phase equilibrium. However, when the phase equilibrium is not set, the degree of freedom of the model is 2 . Moreover,  this SmoothVLE method only calculates phase equilibrium between bubble point and dew point. I think this method may not be suitable for my process.

**Comment by andrewlee94** [comment]
The SmoothVLE method is suitable for your process, however there are a few things you need to understand first.

First and most importantly, IDAES is a fully Equation Oriented based modeling environment, which means that every aspect of the model must be represented using mathematical equations. Further, these equations need to be continuous, at least once-fully differentiable (preferably twice) and feasible at all model states. However, this makes it challenging to represent if/else type statements (such as phase determination) as these are discontinuous by nature.

In practice, this means that we cannot simply do a pre-check to determine what phases are present and react accordingly - we need to solve the equations for all phases at all states. This means that we need to ensure that the equations for all phase are feasible at all states - something which is generally not true for non-existent phases in a VLE system.

The way we get around this is the SmoothVLE method. Rather than solve the VLE at the current system state (where both phase may or may not exist), we instead find the closest state (temperature) where both phases are guaranteed to exist:

* if the current state is in the two-phase envelop, then the current state is used,
* if the current state is fully vapor then the dew point is used, or
* if the current state is fully liquid then the bubble point is used.

We then solve the VLE problem at this new temperature (referred to internally as `_t_eq`). By doing this, we ensure that there is always a feasible solution to the phase equilibrium problem and that it gives a useful solution. If we started in the two-phase region then the problem is solved at the current state and we get the correct answer, and if the current state is in a single phase region then the VLE problem is solved at a state that gives a negligible amount of the non-existent phase (but is still feasible). So, this method is appropriate for you problem.

However, pure component systems complicate this a bit. As a I mentioned earlier, pure component systems have a phase-transition point rather than a two-phase envelope. This means the bubble point is equal to the dew point, and the phase equilibrium is essentially a step change behaviour, Equation oriented solvers have great difficulty dealing with this type of behaviour, although it is possible to solve these types of problems. We have tested the SmoothVLE formulation on single phase systems in the past, and it did work (although it required more effort to solve).
